src/cryoemservices/util/clem_array_functions.py

```diff
@@ -208,6 +214,56 @@ def _by_array_values(array: np.ndarray, dtype_group: str) -> str:
     return dtype_final
 
 
+def shrink_value(value: int) -> int:
+    """
+    When converting between floats and int dtypes in NumPy, there are cases where the
+    float value will be rounded inexactly to be larger than that of its corresponding
+    int, leading to issues when casting arrays with large values.
+
+    This function replaces the value of the integer with one that can be correctly
+    represented in both float and int without a further change in info.
+    """
+
+    # Validate input
+    if not isinstance(value, (int)):
+        raise TypeError(f"Input is not an integer: {str(type(value))}")
+
+    vfloat = float(value)
+
+    # Process number if float conversion increases its value
+    if abs(vfloat) > abs(value):
+        vstr = str(vfloat)
+        if "e" in vstr:
+            # Separate the numerical bit from the exponent
+            num, exp = vstr.split("e", 1)
+
+            # Remove negative sign for subsequent stage
+            if num.startswith("-"):
+                num = num.split("-", 1)[-1]
+                neg = "-"
+            else:
+                neg = ""
+
+            # Reduce precision and shrink value
+            #   NOTE: Python rounds values that are exactly halfway (i.e. 0.5) to the
+            #   nearest EVEN number, so 2 needs to be subtracted to ensure it doesn't
+            #   round up to above the maximum value allowed by the array dtype again.
+            num = str(round((float(num) * 10**15) - 2, 0) / 10**15)
```

Comment:
Does the use of 10**15 set limits on what size numbers this function can handle?